src/Symfony/Component/Translation/Catalogue/ReplaceOperation.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Catalogue;
+
+use Symfony\Component\Translation\MetadataAwareInterface;
+
+/**
+ * This will merge and replace all values in $target with values from $source.
+ * It is the equivalent of running array_merge($target, $source). When in conflict,
+ * always take values from $source.
+ *
+ * This operation is metadata aware. It will do the same recursive merge on metadata.
+ *
+ * all = source ∪ target = {x: x ∈ source ∨ x ∈ target}
+ * new = all ∖ target = {x: x ∈ source ∧ x ∉ target}
+ * obsolete = all ∖ source = {x: x ∈ target ∧ x ∉ source}
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class ReplaceOperation extends AbstractOperation
+{
+    protected function processDomain($domain)
+    {
+        $this->messages[$domain] = array(
+            'all' => array(),
+            'new' => array(),
+            'obsolete' => array(),
+        );
+        $sourceMessages = $this->source->all($domain);
+
+        foreach ($this->target->all($domain) as $id => $message) {
+            $this->messages[$domain]['all'][$id] = $message;
+
+            // If $id is NOT defined in source.
+            if (!array_key_exists($id, $sourceMessages)) {
```

Comment:
What is the reason to mark the message as obsolete then? Doesn't it make more sense to define the set of obsolete message to be the messages from the target catalog that will be replaced by messages from the source catalog?

Reply:
I want to mark what messages that has a definition (translation) but that not exist in source. The reason for this is I might (or most likely) have dynamic translations that cannot be extracted.